File: DailyChallenge/2020/March/16.java
//https://leetcode.com/explore/challenge/card/march-leetcoding-challenge-2021/590/week-3-march-15th-march-21st/3674/

class Solution {
    public int maxProfit(int[] prices, int fee) {
        int notHold = 0, hold = -0x3f3f3f3f;

        for (int i = 0; i < prices.length; i++) {
            hold = Math.max(hold, notHold - prices[i]);
            notHold = Math.max(notHold, hold + prices[i] - fee);
        }

        return notHold;
    }
}

File: DailyChallenge/2020/May/16.cpp
// https://leetcode.com/explore/challenge/card/may-leetcoding-challenge-2021/600/week-3-may-15th-may-21st/3745/

class Solution {
public:
    int minCameraCover(TreeNode* root) {
        int result = 0;
        if (dfs(root, &result) == UNCOVERED) {
            ++result;
        }
        return result;
    }

private:
    int dfs(TreeNode* root, int *result) {
        int left = root->left ? dfs(root->left, result) : COVERED;
        int right = root->right ? dfs(root->right, result) : COVERED;
        if (left == UNCOVERED || right == UNCOVERED) {
            ++(*result);
            return CAMERA;
        }
        if (left == CAMERA || right == CAMERA) {
            return COVERED;
        }
        return UNCOVERED;
    }
    
    enum NODE {
        UNCOVERED, COVERED, CAMERA
    };
};
